src/backend/metal/MetalBackend.mm

```diff
@@ -648,6 +663,99 @@ void EndRenderPass() {
+            case nxt::StencilOperation::Keep:
+                return MTLStencilOperationKeep;
+            case nxt::StencilOperation::Zero:
+                return MTLStencilOperationZero;
+            case nxt::StencilOperation::Replace:
+                return MTLStencilOperationReplace;
+            case nxt::StencilOperation::Invert:
+                return MTLStencilOperationInvert;
+            case nxt::StencilOperation::IncrementClamp:
+                return MTLStencilOperationIncrementClamp;
+            case nxt::StencilOperation::DecrementClamp:
+                return MTLStencilOperationDecrementClamp;
+            case nxt::StencilOperation::IncrementWrap:
+                return MTLStencilOperationIncrementWrap;
+            case nxt::StencilOperation::DecrementWrap:
+                return MTLStencilOperationDecrementWrap;
+        }
+    }
+
+    DepthStencilState::DepthStencilState(Device* device, DepthStencilStateBuilder* builder)
+        : DepthStencilStateBase(builder), device(device) {
+        MTLDepthStencilDescriptor* mtlDepthStencilDescriptor = [MTLDepthStencilDescriptor new];
+
+        if (DepthTestEnabled()) {
+            auto& depth = GetDepth();
+            mtlDepthStencilDescriptor.depthCompareFunction = MetalDepthStencilCompareFunction(depth.compareFunction);
+            mtlDepthStencilDescriptor.depthWriteEnabled = depth.depthWriteEnabled;
+        }
+
+        auto& stencil = GetStencil();
+
+        if (StencilTestEnabled()) {
+            MTLStencilDescriptor* backFaceStencil = [MTLStencilDescriptor new];
+            MTLStencilDescriptor* frontFaceStencil = [MTLStencilDescriptor new];
+
+            backFaceStencil.stencilCompareFunction = MetalDepthStencilCompareFunction(stencil.back.compareFunction);
+            backFaceStencil.stencilFailureOperation = MetalStencilOperation(stencil.back.stencilFail);
+            backFaceStencil.depthFailureOperation = MetalStencilOperation(stencil.back.depthFail);
+            backFaceStencil.depthStencilPassOperation = MetalStencilOperation(stencil.back.depthStencilPass);
+            backFaceStencil.readMask = stencil.back.mask;
+            backFaceStencil.writeMask = stencil.back.mask;
+
+            frontFaceStencil.stencilCompareFunction = MetalDepthStencilCompareFunction(stencil.front.compareFunction);
+            frontFaceStencil.stencilFailureOperation = MetalStencilOperation(stencil.front.stencilFail);
+            frontFaceStencil.depthFailureOperation = MetalStencilOperation(stencil.front.depthFail);
+            frontFaceStencil.depthStencilPassOperation = MetalStencilOperation(stencil.front.depthStencilPass);
+            frontFaceStencil.readMask = stencil.front.mask;
+            frontFaceStencil.writeMask = stencil.front.mask;
+
+            mtlDepthStencilDescriptor.backFaceStencil = backFaceStencil;
+            mtlDepthStencilDescriptor.frontFaceStencil = frontFaceStencil;
+            [backFaceStencil release];
```

Comment:
Checked that this early release doesn't cause a crash, I'm not sure what ObjC magic makes this work.